### PR TITLE
adding a package for working with kubeconfig files

### PR DESCRIPTION
**Pull request by guoshimin**
It's a step towards fixing #2.

Currently it doesn't do much beyond marshaling/unmarshaling kubeconfigs.

### CODE REVIEW

**Comment by guoshimin** [comment]
@greenrd @wayofthepie Can any of you review this PR? Thanks!

File: kubeconfig/src/Kubernetes/KubeConfig.hs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+camelToWithOverrides :: Char -> Map.Map String String -> Options
+camelToWithOverrides c overrides = defaultOptions
+    { fieldLabelModifier = modifier
+    , omitNothingFields  = True
+    }
+    where modifier s = Map.findWithDefault (camelTo2 c s) s overrides
+
+-- |Represents a kubeconfig.
+data Config = Config
+  { kind           :: Maybe (Text)
```
Comment (greenrd) [inline kubeconfig/src/Kubernetes/KubeConfig.hs]: You don't need parentheses here when the type parameter is just a type name.

File: kubeconfig/src/Kubernetes/KubeConfig.hs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        Just auth -> Right (authInfo, auth)
+        Nothing   -> Left ("No user named " <> T.unpack authInfo)
+
+-- |Returns the currently active cluster.
+getCluster :: Config -> Either String Cluster
+getCluster cfg@Config {..} = do
+    Context {..} <- getContext cfg
+    let maybeCluster = Map.lookup cluster (toMap clusters)
+    case maybeCluster of
+        Just cluster -> Right cluster
```
Comment (greenrd) [inline kubeconfig/src/Kubernetes/KubeConfig.hs]: You're shadowing the cluster field here that is being implicitly brought in to scope by the `{..}`, which makes the code confusing to read - it looks like you have a forward reference. Can you rename one of them to remove the shadowing?

File: kubeconfig/test/Spec.hs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+                                        Context (..), getAuthInfo, getCluster,
+                                        getContext)
+import           Test.Hspec
+
+main :: IO ()
+main = do
+  config :: Config <- fromJust <$> decodeFile "test/testdata/kubeconfig.yaml"
+  hspec $ do
+    describe "FromJSON and ToJSON instances" $ do
+      it "roundtrip successfully" $ do
```
Comment (greenrd) [inline kubeconfig/test/Spec.hs]: nit: should be _roundtrips_

File: kubeconfig/src/Kubernetes/KubeConfig.hs (hunk 2 above)
Comment (guoshimin) [reply kubeconfig/src/Kubernetes/KubeConfig.hs]: Done

File: kubeconfig/src/Kubernetes/KubeConfig.hs (hunk 1 above)
Comment (guoshimin) [reply kubeconfig/src/Kubernetes/KubeConfig.hs]: Done

File: kubeconfig/test/Spec.hs (hunk 3 above)
Comment (guoshimin) [reply kubeconfig/test/Spec.hs]: Done

**Comment by greenrd** [comment]
LGTM now.

**Comment by guoshimin** [comment]
@mbohlool @brendandburns Can you guys merge this one? Thanks.

**Comment by guoshimin** [comment]
ping @mbohlool  @brendandburns

**Comment by brendandburns** [comment]
Sorry we should really add one of you as a contributor. @mbohlool what do you think?

**Comment by brendandburns** [comment]
(also feel free to ping me on slack for faster response...)

**Pull request merged**